[TensorPipe] Fix lock inversion upon response read error handling

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* **#38929 [TensorPipe] Fix lock inversion upon response read error handling**
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging



Fixes a TSAN error that was reported by the internal tests.

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703022/)!

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022)

lw committed:
[TensorPipe] Fix lock inversion upon response read error handling

Fixes a TSAN error that was reported by the internal tests.

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703022/)!

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Fix lock inversion upon response read error handling"




Fixes a TSAN error that was reported by the internal tests.

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703022/)!

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022)

[ghstack-poisoned]

mrshenli reviewed (approved):
I can understand that the previous version might be a bit slower as it does looping within the lock scope, but why it was wrong?

osalpekar commented:
> I can understand that the previous version might be a bit slower as it does looping within the lock scope, but why it was wrong?

It was calling setError on the futureMessage with the lock (mutex_) held, so in case any of the callbacks it triggered tried to acquire a lock on mutex_, we would have a deadlock.

lw commented:
I've been trying to find the TSAN error again but didn't manage. From what I remember, the lock inversion error came from an interaction between the timeout thread and the retry thread: the timeout thread would be holding the agent's mutex while completing a future and running its callback, one of which was set by the retry system and was acquiring a mutex there; the retry thread on the other hand would hold the retry mutex while calling the send method, which also acquired the agent's mutex.

lw committed:
Update on "[TensorPipe] Fix lock inversion upon response read error handling"




Fixes a TSAN error that was reported by the internal tests.

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21703022/)!

Differential Revision: [D21703022](https://our.internmc.facebook.com/intern/diff/D21703022)

[ghstack-poisoned]
